kafka-connect-s3/src/main/java/io/confluent/connect/s3/util/S3ErrorUtils.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2021 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.connect.s3.util;
+
+import com.amazonaws.AmazonClientException;
+import com.amazonaws.AmazonWebServiceRequest;
+import com.amazonaws.retry.PredefinedRetryPolicies;
+import io.confluent.connect.storage.common.util.StringUtils;
+import org.apache.commons.lang3.exception.ExceptionUtils;
+import org.apache.kafka.connect.errors.ConnectException;
+import org.apache.kafka.connect.errors.RetriableException;
+
+import java.io.IOException;
+
+/**
+ * Utilities related to basic S3 error/exception analysis.
+ */
+public class S3ErrorUtils {
+
+  /**
+   * Return whether the given exception is a "retryable" exception.
+   * @param exception The exception to analyze
+   * @return true if the exception is retryable
+   */
+  private static boolean isRetriableException(Throwable exception) {
+    if (exception == null) {
+      return false;
+    }
+
+    // IOException, in many places, is passed the AWS exception
+    // when it is thrown. We search here to check that exception
+    // for the IOException case.  Otherwise, the IOException
+    // is considered not retriable.
+    // Exception: An IOException embedded within an `AmazonClientException`
+    // should be passed via the `AmazonClientException` object
+    // as its parent (as the SDK does), in which case, shouldRetry()
+    // will often find it retriable.
+    for (Throwable cause : ExceptionUtils.getThrowableList(exception)) {
+      if (cause instanceof AmazonClientException) {
+        // The AWS SDK maintains a check for what it considers to be
+        // retriable exceptions.
+        return PredefinedRetryPolicies.DEFAULT_RETRY_CONDITION.shouldRetry(
+                AmazonWebServiceRequest.NOOP,
+                (AmazonClientException) cause,
+                Integer.MAX_VALUE
+        );
+      }
+
+      if (!(cause instanceof IOException)) {
```

Comment:
Every exception in the chain would not be IOException. Should we remove the if block as this is an exceptional situation and it's ok to iterate over all of them and outside the loop we are anyway returning false?

Reply:
I'm okay to remove it. However, the logic will be different than in recursive version, originally proposed by @cjolivier01. Please discuss with him offline and let me know your decision.

Reply:
@lion-simba I talked with @cjolivier01 internally and he is ok with removing this. please remove this.